### PR TITLE
[codex] Fix phantom stop propagation

### PR DESCRIPTION
**Pull request by trackrat-dev**
## Summary

Fixes phantom stop propagation across scheduled and live journey update paths.

This PR addresses the backend root cause behind pre-departure Amtrak phantom stops and hardens the related GTFS-RT fuzzy-trip update behavior across the MTA-family collectors. It also converts NJT phantom-stop cleanup to the safer bulk-delete pattern and keeps `stops_count` aligned with the current stop collection.

## Root Cause

Amtrak scheduled journeys were seeded by copying the full stop list from a single recent `OBSERVED` journey and shifting timestamps. If that one observed run included an unusual one-off stop, every future `SCHEDULED` record for that train number inherited the phantom stop until live Amtrak reconciliation corrected it.

The MTA-family collectors had a latent version of the same orphan-stop risk: their exact train-id updates are safe, but their fuzzy trip-id fallback could accept a different live trip with a different stop set and then leave old stops attached to the existing journey.

The follow-up test failure exposed a mixed station-code edge case in the new Amtrak consensus guard: patterns can carry raw Amtrak codes such as `WAS`, while journey stops may carry TrackRat internal codes such as `WS`. The required origin/terminal check now accepts either raw or mapped internal aliases.

PR feedback also caught that alias handling must happen before consensus construction: the recent observed-journey query itself was still filtering route samples with strict origin/terminal equality, so alias-equivalent historical rows could be dropped before they reached the consensus step.

A later review pass found two remaining edge cases: MTA-family fuzzy candidates were being pre-filtered before stop-set comparison, hiding extra candidate stops, and Amtrak stop consensus still compared raw/internal stop aliases as distinct stations.

### Amtrak pattern scheduler

- Replaced single-run scheduled stop seeding with recent same-weekday, same-route consensus.
- Requires at least two complete observed samples before trusting a scheduled route stop list.
- Intersects canonicalized Amtrak station codes across sample runs so one-off stops are excluded while raw/internal aliases still agree.
- Queries recent route samples using raw/internal Amtrak origin and terminal aliases.
- Requires origin and terminal to be present in the consensus before using it, while tolerating raw/internal Amtrak station-code aliases.
- Falls back to the origin-only scheduled stop when a stable route cannot be proven.
- Uses median stop sequence and median time offsets to build scheduled stop templates.
- Normalizes stop sequences before saving scheduled rows.

### MTA-family collectors

- Added shared `select_matching_trip` helper for LIRR, MNR, Subway, BART, MBTA, and Metra.
- Added shared candidate grouping that preserves full live trips before fuzzy stop-set comparison.
- Preserves exact train-id matching behavior, including partial live-feed stop sets, while applying exact updates only to stored journey stops.
- Makes fuzzy fallback stricter by rejecting candidates whose full visible station set differs from the stored journey.
- Keeps `journey.stops_count` refreshed whenever callers have the current stop list.

### NJT journey collector

- Replaced select-then-per-row phantom stop deletion with a single bulk `DELETE` scoped to stops not present in the live API response.
- Leaves existing resequencing and duplicate-sequence validation in place after cleanup.

### Tests

- Updated Amtrak scheduler tests to cover consensus-built scheduled stops.
- Added coverage that a one-off Amtrak stop is excluded from scheduled rows.
- Added coverage that the recent observed-journey query includes alias-equivalent Amtrak origin/terminal station codes.
- Added coverage that raw/internal Amtrak stop aliases intersect during consensus.
- Added MTA common tests for exact match tolerance, exact match scoping, fuzzy mismatch rejection, fuzzy extra-stop rejection, fuzzy same-stop-set acceptance, candidate grouping, and `stops_count` refresh.

## Safety Notes

- The Amtrak change is intentionally conservative: if there is not enough agreement among recent complete observed runs, scheduled journeys keep only the origin stop instead of inventing a full route.
- The route-sample query, required-stop check, and stop-consensus intersection now handle mixed raw/internal Amtrak station codes like `WAS`/`WS`.
- Exact GTFS-RT trip matches remain permissive for partial feeds, so normal updates are not blocked when passed stops disappear from live data.
- Exact GTFS-RT update application remains scoped to stops that already exist on the stored journey.
- Fuzzy GTFS-RT matches now only update an existing journey when the full visible stop set is the same, avoiding cross-variant mutation.
- NJT still resequences and validates after deleting phantom stops.

## Validation

- `PYTHONPYCACHEPREFIX=/private/tmp/trackrat_pycache /Users/andy/.cache/codex-runtimes/codex-primary-runtime/dependencies/python/bin/python3 -m py_compile ...` on all modified source and test files.
- `git diff --check`.
- Focused pytest could not be run locally because this environment does not have `pytest`, `poetry`, or `uv` installed.

## Commits

`f56f0bf Fix phantom stop propagation`

- Build Amtrak scheduled stop lists from recent same-route consensus instead of copying a single observed run, with origin-only fallback when the stable route cannot be proven.
- Centralize GTFS-RT exact/fuzzy trip selection so MTA-family collectors reject fuzzy matches whose stop sets differ from the stored journey, while keeping exact train-id matches tolerant of partial live feeds.
- Refresh `stops_count` from the current stop collection and convert NJT phantom-stop cleanup to a bulk delete before resequencing.
- Validation: py_compile on modified backend/test files; git diff --check. Pytest intentionally skipped per request.

`9b42045 Handle Amtrak station aliases in stop consensus`

- Allow required scheduler origin and terminal codes to match either raw Amtrak station codes or mapped internal TrackRat station codes.
- Keeps the consensus safety check intact while fixing mixed-code historical data such as `WAS` versus `WS`.
- Validation: py_compile for the scheduler and related test file; git diff --check. Focused pytest could not run locally because pytest/poetry/uv are not installed in this environment.

`9113bea Query Amtrak route samples by station aliases`

- Expand the recent observed-journey query to match raw Amtrak and internal TrackRat origin/terminal aliases before building scheduled-stop consensus.
- Addresses PR feedback where strict `WAS` versus `WS` route filters could drop otherwise valid historical samples and force origin-only scheduled journeys.
- Validation: py_compile for the scheduler and related scheduler test file; git diff --check.

`a85c2b8 Tighten stop-set reconciliation edge cases`

- Group full GTFS-RT candidate trips before fuzzy matching so added route-variant stops remain visible to the stop-set comparison.
- Keep exact train-id updates scoped to stored journey stops to preserve existing update behavior while still tolerating partial live feeds.
- Canonicalize Amtrak raw station codes to TrackRat internal codes before stop-consensus intersection so mixed raw/internal historical stops still agree.
- Validation: py_compile for modified backend and test files; git diff --check.

### CODE REVIEW

File: backend_v2/src/trackrat/services/amtrak_pattern_scheduler.py [hunk 1]
```diff
@@ -374,43 +376,177 @@ async def analyze_historical_patterns_optimized(
...
             .where(
                 and_(
-                    TrainJourney.train_id == train_number,
+                    TrainJourney.train_id == pattern["train_number"],
                     TrainJourney.data_source == "AMTRAK",
                     TrainJourney.observation_type == "OBSERVED",
                     TrainJourney.has_complete_journey.is_(True),
+                    TrainJourney.is_cancelled.is_(False),
+                    TrainJourney.origin_station_code == pattern["origin"],
+                    TrainJourney.terminal_station_code == pattern["terminal"],
```
[the remark this answers is not in the document]
Comment (trackrat-dev) [reply backend_v2/src/trackrat/services/amtrak_pattern_scheduler.py]: This was addressed in commit `9113bea` ("Query Amtrak route samples by station aliases"), which expanded the recent observed-journey query to match both raw Amtrak and internal TrackRat origin/terminal aliases.

That said, after tracing the code more deeply, this concern may be moot in practice: the Amtrak journey collector (`collectors/amtrak/journey.py:221`) normalizes all raw Amtrak codes through `AMTRAK_TO_INTERNAL_STATION_MAP` before storage — `WAS` → `WS`, `NYP` → `NY`, etc. So `origin_station_code` and `terminal_station_code` in `train_journeys` for OBSERVED Amtrak records should always contain internal codes, never raw ones. The alias expansion in `9113bea` is a defensive belt-and-suspenders measure, which is fine to keep.

One related observation: the test file `test_amtrak_pattern_scheduler.py` uses raw Amtrak codes (`NYP`, `WAS`, `NWK`, `TRE`) in mock data instead of the internal codes (`NY`, `WS`, `NP`, `TR`) that would actually appear in the database. Not a runtime bug since the tests mock the DB layer, but the mock data doesn't reflect production reality.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by trackrat-dev** [comment]
## High-Level Review Overview

### Codex Review Comment — Station Alias Concern
The single review comment from Codex flagged that strict `origin_station_code`/`terminal_station_code` filtering in the consensus query could drop alias-equivalent rows (`WAS` vs `WS`). This was addressed in commit `9113bea`, which expanded the query to match both raw Amtrak and internal station code aliases.

After tracing the full code path, the concern is actually moot in production: the Amtrak journey collector normalizes all raw codes to internal codes via `AMTRAK_TO_INTERNAL_STATION_MAP` before storage, so OBSERVED records should never have raw codes in origin/terminal fields. The alias expansion is a harmless defensive measure.

### CI Status
All 5 checks pass: Backend Tests, Docker Build, Terraform Validation, Webpage Build, CI Summary.

### Minor Observation
Test mock data in `test_amtrak_pattern_scheduler.py` uses raw Amtrak codes (`NYP`, `WAS`) instead of the internal codes that production data would contain (`NY`, `WS`). Not a blocker — the tests mock the DB layer so they still validate the logic correctly — but worth noting for future test accuracy.

### Recommendation
**Ready to merge.** The Codex concern has been addressed, all CI checks pass, and the overall approach (consensus-based stop seeding, stricter fuzzy matching, bulk phantom-stop deletion) is sound.

---
_Generated by [Claude Code](https://claude.ai/code)_
